### PR TITLE
[Tailcall] Give LLVM more leeway.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -7046,7 +7046,8 @@ is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmetho
 		|| (cmethod->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL)
 		|| cfg->method->save_lmf
 		|| (cmethod->wrapper_type && cmethod->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD)
-		|| call_opcode == CEE_CALLI
+		|| (!cfg->llvm_only && (
+		   call_opcode == CEE_CALLI
 		|| (virtual_ && !cfg->backend->have_op_tail_call_membase)
```
Comment (vargaz) [inline mono/mini/method-to-ir.c]: This if looks pretty complicated now.

Comment (jaykrell) [reply mono/mini/method-to-ir.c]: There is a slight restructuring that might help you read it. Let me try.

**Comment by jaykrell** [comment]
Better? This is a bit experimental as I haven't done the test but I think it is reasonable.
In either case, this makes sense to separate from the other PR in which it first appeared.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Interpreter

**Comment by jaykrell** [comment]
It should have slightly less, and it should/could decide for itself.
In particular, it should have fewer restrictions around generic context and maybe interface method dispatch. i.e. should enable ARM32 to tailcall more, with LLVM.

**Comment by jaykrell** [comment]
As I understand, llvmonly has at least one fewer restrictions than mini and we should let it take advantage that.
Specifically that rgctx/imt is passed “within” the ABI and not “outside” of the ABI.
That should let more things work with llvmonly — at least with ARM32.
It might have advantages in gsharedvt also.

I feels like a lot of our checks are about mini, and LLVM should just be allowed to do whatever it can, at least given the tail. prefix — maybe even without it. Even things like lmf it should be able to figure out itself cannot be tailcalled, because there is stuff after the call. I think we are just unnecessarily tyings its hands.

Other examples are the stack_usage==0 in arm64, or stack_usage < 64 in arm32. I’d like to visit these, understand them, see about fixing them, but perhaps they can be lifted right away for llvmonly.

**Comment by jaykrell** [comment]
Maybe I add some tests that this moves from failing to passing? I believe I can.

**Comment by vargaz** [comment]
Even if llvm supports this, I don't think we should support different tail calls with different configurations, it would mean that an app would work in some config and not in others. Also, llvmonly is only used on the watch, and any potential problems are hard to debug there.

**Comment by jaykrell** [comment]
Right, I wasn't sure about that -- @luhenry is the goal to fill as many spots in the matrix, or to have it be kinda "level".

Will any F# user be serious enough for about arm32 tailcall so as to use llvmonly because it can be better?

And what are we willing to do for arm32? For example, generate llvm-like generic thunks instead of using the extra register? Or just don't do arm32 well?

**Comment by jaykrell** [comment]
Ps: We can certainly enable llvmonly CI on more platforms, independently of the shipping scenarios. But that doesn't address the kinda jaggy/sparse resulting matrix which I think is the larger concern.

**Comment by jaykrell** [comment]
The matrix: https://github.com/mono/mono/issues/6914

**Comment by luhenry** [comment]
@jaykrell we want to strive to have the same features working across the board, with the expected slight differences between platforms. Any difference needs to be justified, limited and understood; this is to minimize chances of a bug appearing in 1 configuration and not the other, thus for obvious reasons of limited manpower and the load of maintaining all these configurations. Unfortunately, giving LLVM more leeway is not limited nor justified, since it will give to the `bitcode`/`llvmonly` configuration different constraints than to the others.

**Pull request closed without merging**